x/batch/keeper/delegation.go

```diff
@@ -0,0 +1,130 @@
+	// period has already been incremented - we want to store the period ended by this delegation action
+	previousPeriod := k.distrKeeper.GetValidatorCurrentRewards(ctx, val).Period - 1
+
+	// increment reference count for the period we're going to track
+	k.incrementReferenceCount(ctx, val, previousPeriod)
+
+	validator := k.stakingKeeper.Validator(ctx, val)
+	delegation := k.stakingKeeper.Delegation(ctx, del, val)
+
+	// calculate delegation stake in tokens
+	// we don't store directly, so multiply delegation shares * (tokens per share)
+	// note: necessary to truncate so we don't allow withdrawing more rewards than owed
+	stake := validator.TokensFromSharesTruncated(delegation.GetShares())
+	k.distrKeeper.SetDelegatorStartingInfo(ctx, val, del, types.NewDelegatorStartingInfo(previousPeriod, stake, uint64(ctx.BlockHeight())))
+}
+
+func (k Keeper) withdrawAllDelegationRewards(ctx sdk.Context, delAddr sdk.AccAddress) (sdk.Coins, error) {
+	rewardsTotal := sdk.Coins{}
+	remainderTotal := sdk.DecCoins{}
+	k.stakingKeeper.IterateDelegations(ctx, delAddr, func(_ int64, del stakingtypes.DelegationI) (stop bool) {
+		valAddr := del.GetValidatorAddr()
+		val := k.stakingKeeper.Validator(ctx, valAddr)
+
+		// check existence of delegator starting info
+		if !k.distrKeeper.HasDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr()) {
+			return false
+		}
+
+		// end current period and calculate rewards
+		endingPeriod := k.distrKeeper.IncrementValidatorPeriod(ctx, val)
+		rewardsRaw := k.distrKeeper.CalculateDelegationRewards(ctx, val, del, endingPeriod)
+		outstanding := k.distrKeeper.GetValidatorOutstandingRewardsCoins(ctx, del.GetValidatorAddr())
+
+		// defensive edge case may happen on the very final digits
+		// of the decCoins due to operation order of the distribution mechanism.
+		rewards := rewardsRaw.Intersect(outstanding)
+		if !rewards.IsEqual(rewardsRaw) {
+			logger := k.Logger(ctx)
+			logger.Info(
+				"rounding error withdrawing rewards from validator",
+				"delegator", del.GetDelegatorAddr().String(),
+				"validator", val.GetOperator().String(),
+				"got", rewards.String(),
+				"expected", rewardsRaw.String(),
+			)
+		}
+
+		// truncate reward dec coins, return remainder to community pool
+		finalRewards, remainder := rewards.TruncateDecimal()
+		remainderTotal = remainderTotal.Add(remainder...)
+		rewardsTotal = rewardsTotal.Add(finalRewards...)
+
+		// update the outstanding rewards and the community pool only if the
+		// transaction was successful
+		k.distrKeeper.SetValidatorOutstandingRewards(ctx, del.GetValidatorAddr(), types.ValidatorOutstandingRewards{Rewards: outstanding.Sub(rewards)})
+
+		// decrement reference count of starting period
+		startingInfo := k.distrKeeper.GetDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())
+		startingPeriod := startingInfo.PreviousPeriod
+		k.decrementReferenceCount(ctx, del.GetValidatorAddr(), startingPeriod)
+
```

Comment:
x/distribution has `k.DeleteDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())` here - do we need it?

Reply:
no, we don't need it because `DelegatorStartingInfo` is created again in `initializeDelegation` function.
https://github.com/cosmos/cosmos-sdk/blob/main/x/distribution/keeper/keeper.go#L204